include/urbi/object/centralized-slots.hxx

```diff
@@ -42,8 +42,36 @@ namespace urbi
     inline void
     CentralizedSlots::finalize(Object* owner)
     {
+#if BOOST_VERSION <= 105500
       if (owner->slots_.size_)
         obj_index_.erase(owner);
+#else
+      // boost::multi_index_container with boost > 1.55 does not seem to support reentrency into erase,
+      // so we remove this reentrency by appending objects to finalize into delayed_finalize_.
+      // delayed_finalize_ is immediately handled so this should not have any impact on the language semantic.
+      if (owner->slots_.size_)
+      {
+        if (in_finalize_)
+        {
+          delayed_finalize_.push_back(owner);
+          return;
+        }
+        {
+          libport::Finally finally;
+          finally << libport::scoped_set(in_finalize_, true);
+          obj_index_.erase(owner);
+        }
+        while (!delayed_finalize_.empty())
+        {
+          std::vector<Object*> copy = delayed_finalize_;
+          delayed_finalize_.clear();
+          for (auto o: copy)
```

Comment:
and adds a lambda or a bind, I'm not convinced.

Reply:
Well, I still think avoiding loops is nice.  Conversely, FWIW, this would be a nice use case for C++20's for-loop initializers :)
```
for (auto obj = std::move(delayed_finalize_); auto o: obj)
  ...
```

Reply:
Or possibly C++14's std::exchange.

```
for (auto o: std::exchange(obj, vector<Object*>{}))
```